File: notebooks/gather_trisk_results.py
import pandas as pd
import os
from pathlib import Path
import glob


def gather_trisk_results(artifacts_dir: str) -> pd.DataFrame:
    """
    Gather TRISK NPV results from downloaded artifacts.

    Args:
        artifacts_dir (str): Directory containing the downloaded artifacts

    Returns:
        pd.DataFrame: Combined NPVs DataFrame
    """
    # Get all NPV files
    npv_files = glob.glob(os.path.join(artifacts_dir, "npvs_*.csv"))

    # Initialize empty DataFrame for combined results
    all_npvs = pd.DataFrame()

    # Process NPV files
    for file in npv_files:
        run_id = os.path.basename(file).replace("npvs_", "").replace(".csv", "")
        df = pd.read_csv(file)
        df["run_id"] = run_id
        all_npvs = pd.concat([all_npvs, df], ignore_index=True)

    return all_npvs


def split_by_ccs_status(
    df: pd.DataFrame, ccs_technologies: list[str] = ["GasCap", "CoalCap", "OilCap"]
) -> dict[str, pd.DataFrame]:
    """
    Split the DataFrame into three groups based on CCS status.

    Args:
        df (pd.DataFrame): Input DataFrame with technology column
        ccs_technologies (list[str], optional): List of technologies that can have CCS.
            Defaults to ['GasCap', 'CoalCap', 'OilCap']

    Returns:
        dict[str, pd.DataFrame]: Dictionary containing three filtered DataFrames:
            - 'with_ccs': Technologies with CCS + other technologies
            - 'without_ccs': Technologies without CCS + other technologies
            - 'other': All technologies except CCS-capable ones
    """
    # Get unique run IDs for each CCS variant
    with_ccs_runs = df[df["technology"].str.contains("w/ CCS", na=False)][
        "run_id"
    ].unique()
    without_ccs_runs = df[df["technology"].str.contains("w/o CCS", na=False)][
        "run_id"
    ].unique()

    # Split technologies by CCS status using run IDs
    with_ccs = df[df["run_id"].isin(with_ccs_runs)]
    without_ccs = df[df["run_id"].isin(without_ccs_runs)]

    # For other technologies, we need to ensure we don't duplicate them
    # We'll take them from the run that has the most complete data
    other_runs = set(df["run_id"].unique()) - set(with_ccs_runs) - set(without_ccs_runs)
    if other_runs:
        other_techs = df[df["run_id"].isin(other_runs)]

    return {"with_ccs": with_ccs, "without_ccs": without_ccs, "other": other_techs}


def save_trisk_results(df: pd.DataFrame, output_dir: str, filename: str):
    """
    Save TRISK NPV results to CSV file.

    Args:
        df (pd.DataFrame): DataFrame to save
        output_dir (str): Directory to save the results
        filename (str): Name of the output file
    """
    # Create output directory if it doesn't exist
    os.makedirs(output_dir, exist_ok=True)

    # Save results
    output_path = os.path.join(output_dir, filename)
    print(f"\nSaving {filename} to: {output_path}")
    df.to_csv(output_path, index=False)
    print(f"Total rows: {len(df)}")


if __name__ == "__main__":
    # Assuming the artifacts are in workspace/mlflow_results/artifacts
    artifacts_dir = "workspace/mlflow_results/artifacts"
    # Save one level up from artifacts directory
    output_dir = "workspace/mlflow_results"

    # Define CCS technologies
    ccs_technologies = ["GasCap", "CoalCap", "OilCap"]

    # Gather the results
    npvs_df = gather_trisk_results(artifacts_dir)

    # Split the results by CCS status
    ccs_split = split_by_ccs_status(npvs_df, ccs_technologies)

    # Save each split
    for ccs_type, df in ccs_split.items():
        save_trisk_results(df, output_dir, f"npvs_{ccs_type}.csv")

File: notebooks/gather_trisk_results.R
library(tidyverse)

#' Gather TRISK NPV results from downloaded artifacts.
#'
#' @param artifacts_dir Character. Directory containing the downloaded artifacts
#' @return Data frame. Combined NPVs data frame
gather_trisk_results <- function(artifacts_dir) {
  # Get all NPV files
  npv_files <- list.files(
    path = artifacts_dir,
    pattern = "npvs_.*\\.csv$",
    full.names = TRUE
  )
  
  # Initialize empty list for results
  all_npvs_list <- list()
  
  # Process NPV files
  for (file in npv_files) {
    run_id <- gsub("npvs_|\\.csv$", "", basename(file))
    df <- read.csv(file)
    df$run_id <- run_id
    all_npvs_list[[length(all_npvs_list) + 1]] <- df
  }
  
  # Combine all data frames
  all_npvs <- bind_rows(all_npvs_list)
  return(all_npvs)
}

#' Split the data frame into three groups based on CCS status.
#'
#' @param df Data frame. Input data frame with technology column
#' @param ccs_technologies Character vector. List of technologies that can have CCS.
#'   Defaults to c('GasCap', 'CoalCap', 'OilCap')
#' @return List. List containing three filtered data frames:
#'   - 'with_ccs': Technologies with CCS + other technologies
#'   - 'without_ccs': Technologies without CCS + other technologies
#'   - 'other': All technologies except CCS-capable ones
split_by_ccs_status <- function(df, ccs_technologies = c("GasCap", "CoalCap", "OilCap")) {
  # Get unique run IDs for each CCS variant
  with_ccs_runs <- unique(df[grepl("w/ CCS", df$technology), "run_id"])
  without_ccs_runs <- unique(df[grepl("w/o CCS", df$technology), "run_id"])
  
  # Split technologies by CCS status using run IDs
  with_ccs <- df[df$run_id %in% with_ccs_runs, ]
  without_ccs <- df[df$run_id %in% without_ccs_runs, ]
  
  # For other technologies, we need to ensure we don't duplicate them
  # We'll take them from the run that has the most complete data
  other_runs <- setdiff(unique(df$run_id), union(with_ccs_runs, without_ccs_runs))
  if (length(other_runs) > 0) {
    other_techs <- df[df$run_id %in% other_runs, ]
  }
  
  return(list(
    with_ccs = with_ccs,
    without_ccs = without_ccs,
    other = other_techs
  ))
}

#' Save TRISK NPV results to CSV file.
#'
#' @param df Data frame. Data frame to save
#' @param output_dir Character. Directory to save the results
#' @param filename Character. Name of the output file
save_trisk_results <- function(df, output_dir, filename) {
  # Create output directory if it doesn't exist
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)
  
  # Save results
  output_path <- file.path(output_dir, filename)
  cat(sprintf("\nSaving %s to: %s\n", filename, output_path))
  write.csv(df, output_path, row.names = FALSE)
  cat(sprintf("Total rows: %d\n", nrow(df)))
}

# Main execution
if (interactive()) {
  # Assuming the artifacts are in workspace/mlflow_results/artifacts
  artifacts_dir <- "workspace/mlflow_results/artifacts"
  # Save one level up from artifacts directory
  output_dir <- "workspace/mlflow_results"
  
  # Define CCS technologies
  ccs_technologies <- c("GasCap", "CoalCap", "OilCap")
  
  # Gather the results
  npvs_df <- gather_trisk_results(artifacts_dir)
  
  # Split the results by CCS status
  ccs_split <- split_by_ccs_status(npvs_df, ccs_technologies)
  
  # Save each split
  for (ccs_type in names(ccs_split)) {
    save_trisk_results(
      ccs_split[[ccs_type]],
      output_dir,
      sprintf("npvs_%s.csv", ccs_type)
    )
  }
} 